### DISCUSSION TITLE
Change startup order of job -> services to services -> job

### DISCUSSION THREAD

**Discussion by umarcor** [community/community#9056] [Actions]
Please, change the startup order of GitHub Actions job from:

1. Starting job container
2. Starting service container
3. Waiting for all services to be ready

to:

1. Starting service container
2. Waiting for all services to be ready
3. Starting job container

That would allow to use services for doing modifications on the host before running the job container. See further details in actions/virtual-environments#2095.

**Comment by AlexTMjugador** [comment]
This seemingly simple change would allow compiling C/C++/Rust projects that depend on C/C++ code with native toolchains and packages, which saves quite a bit of headaches when cross-compilation toolchains or pkg-config definition files are not that feasible, while keeping all the build steps nicely defined inside the workflow file, and not needing an additional container. For context, these are the kind of ugly hacks that people are doing right now: https://gitlab.com/rust_musl_docker/image/-/tree/master

**Comment by ethomson** [comment]
Sorry, I read through https://github.com/actions/virtual-environments/issues/2095 and I don't think that I fully understand the _why_ here yet.  It sounds like your fundamental concern is that you want to be able to do something like `docker run --rm --privileged multiarch/qemu-user-static:register --reset` before you run a job container.  Is that roughly accurate?  You're just trying to use service containers as a means to get there?

**Reply by umarcor** [reply]
@ethomson, yes. The point is to be able to setup services which need to be done before the container job is created. The purpose is not only to reset QEMU interpreters, but to register additional interpreters.

**Reply by umarcor** [reply]
See, for instance, this demo: https://github.com/dbhi/qus/commit/e91758940dfe2b3054ea3b79d11e4dd47918ce1a

**Reply by ethomson** [reply]
Thanks, I just wanted to make sure that I understood.
